### PR TITLE
[ENH] Implementation of LOF using PyODAdapter

### LINKED ISSUE #2109: [ENH] Add PyODAdapter-implementation for LOF
### Describe the feature or idea you want to propose

The [`PyODAdapter`](https://github.com/aeon-toolkit/aeon/blob/main/aeon/anomaly_detection/_pyodadapter.py) in aeon allows us to use any outlier detector from [PyOD](https://github.com/yzhao062/pyod), which were originally proposed for relational data, also for time series anomaly detection (TSAD). Not all detectors are equally well suited for TSAD, however. We want to represent the frequently used and competitive outlier detection techniques within the `anomaly_detection` module of aeon directly.

**Implement the [LOF method](https://github.com/yzhao062/pyod/blob/master/pyod/models/lof.py#L19)** using the `PyODAdapter`.

### Describe your proposed solution

- Create a new file in `aeon.anomaly_detection` for the method
- Create a new estimator class with `PyODAdapter` as the parent
- Expose the algorithm's hyperparameters as constructor arguments, create the PyOD model and pass it to the super-constructor
- Document your class
- Add tests for certain edge cases if necessary

---

Example for IsolationForest:

```python
class IsolationForest(PyODAdapter):
    """documentation ..."""
    def __init__(n_estimators: int = 100, max_samples: int | str = "auto", ..., window_size: int, stride: int):
        model = IForest(n_estimators, max_samples, ...
        super().__init__(model, window_size, stride)

    @classmethod
    def get_test_params(cls, parameter_set="default"):
        """..."""
        return {"n_estimators": 10, ...}
```

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by aryanpola**
#### Reference Issues/PRs
Fixes #2109 

#### What does this implement/fix? Explain your changes.

Implemented Local Outlier Factor using PyODadapter

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

### CODE REVIEW

**Comment by aryanpola** [comment]
@CodeLionX some tests ran with no `window_size` parameter causing errors, would it be a good idea to set the default window size if not provided, to be the same as that of stride?

File: aeon/anomaly_detection/_lof.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+import numpy as np
+from sklearn.exceptions import NotFittedError
+
+from aeon.anomaly_detection._pyodadapter import PyODAdapter
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+
+class LOF(PyODAdapter):
+    """Local Outlier Factor (LOF) algorithm anomaly detection.
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:16]:
```suggestion
    """Local Outlier Factor (LOF) algorithm for anomaly detection.
```

File: aeon/anomaly_detection/_lof.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        distance between them. This works for Scipy's metrics, but is less
+        efficient than passing the metric name as a string.
+    p : integer, optional (default = 2)
+        Parameter for the Minkowski metric
+    metric_params : dict, optional (default = None)
+        Additional keyword arguments for the metric function.
+    contamination : float in (0., 0.5), optional (default=0.1)
+        The amount of contamination of the data set, i.e. the proportion
+        of outliers in the data set. When fitting this is used to define the
+        threshold on the decision function.
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:58]: This parameter is just used to compute a threshold to predict the labels. We do not use this and return the anomaly scores. It can be safely removed (use the PyOD default).

File: aeon/anomaly_detection/_lof.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+from aeon.anomaly_detection._pyodadapter import PyODAdapter
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+
+class LOF(PyODAdapter):
+    """Local Outlier Factor (LOF) algorithm anomaly detection.
+
+    This class implement metrics-based outlier detection algorithms using the
+    Local Outlier Factor (LOF) algorithm from PyOD.
+
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:20]: Please add the capability tables (see other anomaly detectors as an example)

File: aeon/anomaly_detection/_lof.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    n_jobs : int, optional (default = 1)
+        The number of parallel jobs to run for neighbors search.
+        If ``-1``, then the number of jobs is set to the number of CPU cores.
+        Affects only kneighbors and kneighbors_graph methods.
+    novelty : bool (default=False)
+        By default, LocalOutlierFactor is only meant to be used for outlier
+        detection (novelty=False). Set novelty to True if you want to use
+        LocalOutlierFactor for novelty detection. In this case be aware that
+        that you should only use predict, decision_function and score_samples
+        on new unseen data and not on the training set.
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:68]: This is important for the correct usage of the detector. However, I would not expose this parameter to the user and handle it internally according to our usage styles:

We want to support both learning types: semi-supervised and unsupervised. We can distinguish between those because different functions are called:

- Unsupervised

  ```python
  lof = LOF()
  scores = lof.fit_predict(X)
  ```
  
  In this case, we need to set `novelty=False` in the internal model.

- Semi-supervised

  ```python
  lof = LOF()
  lof.fit(X_train)
  scores = lof.predict(X_test)
  ```
  
  In this case, we need to set `novelty=True` in the internal model.

Please override the `novelty`-parameter in the corresponding functions before using the internal model, and remove the parameter from the init.
Example for `_fit_predict()`:

```python
def _fit_predict(self, X: np.ndarray, y: np.ndarray | None = None) -> np.ndarray:
    self.pyod_model.novelty = False
    super()._fit_predict(X, y)
```

File: aeon/anomaly_detection/_lof.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        stride: int = 1,
+    ):
+        _check_soft_dependencies(*self._tags["python_dependencies"])
+        from pyod.models.lof import LOF
+
+        # Validate that stride is not greater than winow_size
+        if stride > window_size:
+            raise ValueError(
+                f"Stride ({stride}) cannot be greater than window size ({window_size})."
+            )
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:100]: Parameter checking should not be implemented in the `__init__` method according to sklearn-conventions. In addition, the base class already provides this for you:

https://github.com/aeon-toolkit/aeon/blob/baed5324ed447ac69f6b2f4b5acf62882b99d3fb/aeon/anomaly_detection/_pyodadapter.py#L134-L147

This is automatically called if you forward the function calls correctly using `super()`.

File: aeon/anomaly_detection/_lof.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        if not hasattr(self, "fitted_pyod_model_"):
+            raise NotFittedError(
+                "This instance of LOF has not been fitted yet; please call `fit` first."
+            )
+        return super()._predict(X)
+
+    def fit_predict(
+        self, X: np.ndarray, y: Union[np.ndarray, None] = None
+    ) -> np.ndarray:
+        return super()._fit_predict(X, y)
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:143]: Do not override `fit`, `predict`, and `fit_predict`! They contain important input validation and transformation logic from the base classes. They are also declared final; it is unclear to me how Python allows overwriting them 🤷🏼 

Please just override the internal methods `_fit`, `_predict`, and `_fit_predict`

File: aeon/anomaly_detection/_lof.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+    def _check_params(self, X: np.ndarray) -> None:
+        if self.window_size < 1 or self.window_size > X.shape[0]:
+            self.window_size = min(max(1, self.window_size), X.shape[0])
+            raise ValueError(
+                "The window size must be at least 1 and "
+                "at most the length of the time series."
+            )
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:151]: Please remove. The `PyODAdapter` already contains this method, but has more checks.

File: aeon/anomaly_detection/_lof.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    leaf_size : int, optional (default=30)
+        Leaf size passed to `BallTree` or `KDTree`. This can
+        affect the speed of the construction and query, as well as the memory
+        required to store the tree. The optimal value depends on the
+        nature of the problem.
+    metric : string or callable, default 'minkowski'
+        metric used for the distance computation. Any metric from scikit-learn
+        or scipy.spatial.distance can be used.
+        If 'precomputed', the training input X is expected to be a distance
+        matrix.
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:45]: We do not support `precomputed` within aeon.

**Review by CodeLionX — CHANGES_REQUESTED** [review]
Thanks for your contribution! The `window_size` and `stride` parameters should have a proper default that should get picked up during the tests. I cannot explain why this is not happening in your tests.

This already looks good. However, LOF is special because of its different calling conventions for unsupervised (outlier detection) and semi-supervised (novelty detection) problems. Please refer to my comment below on how to best integrate this into our API.

The tests in their current version are not very helpful. They mostly test the functionality of the `PyODAdapter`. Similar to the tests in COPOD (https://github.com/aeon-toolkit/aeon/pull/2202#issuecomment-2416920728), I would like to test that the integration into aeon works as expected. We should also test that the `metric`-parameter works for our distance implementations and that the different calling conventions (unsupervised and semi-supervised) work correctly.

**Comment by aryanpola** [comment]
@CodeLionX The output values from aeon_lof and pyod_lof don't match. I've tried normalizing the series but the error persists.
```
FAILED aeon/anomaly_detection/tests/test_lof.py::test_lof_compare_with_pyod_direct - AssertionError: 
Not equal to tolerance rtol=1e-05, atol=0

Mismatched elements: 80 / 80 (100%)
Max absolute difference: 0.11725756
Max relative difference: 0.06374909
 x: array([1.516172, 1.018181, 1.222005, 1.838367, 1.574194, 1.062766,
       1.206536, 0.994702, 0.994427, 1.020478, 0.952964, 1.365051,
       1.122292, 0.956126, 1.028726, 1.004674, 1.378998, 0.982227,...
 y: array([1.433783, 1.000109, 1.184898, 1.728743, 1.486901, 1.06169 ,
       1.170488, 0.986849, 0.976589, 1.001037, 0.951811, 1.295437,
       1.097627, 0.952606, 1.020956, 1.000109, 1.308204, 0.981894,...
```

File: aeon/anomaly_detection/tests/test_lof.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
+"""Tests for the LOF class."""
+
+import numpy as np
+import pytest
+
+from aeon.anomaly_detection import LOF
+from aeon.testing.data_generation._legacy import make_series
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:7]: Please consider using `from aeon.testing.data_generation import make_example_1d_numpy` instead of the legacy testing methods

File: aeon/anomaly_detection/tests/test_lof.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    lof_aeon = LOF(window_size=1, stride=1)
+    pred_aeon = lof_aeon.fit_predict(series)
+
+    # Directly using PyOD LOF
+    lof_pyod = PyOD_LOF()
+    lof_pyod.fit(series.reshape(-1, 1))
+    pred_pyod = lof_pyod.decision_function(series.reshape(-1, 1))
+
+    # Checking if the predictions match
+    np.testing.assert_allclose(pred_aeon, pred_pyod, rtol=1e-5)
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:83]: This mixes calling conventions:

- you use `fit_predict` for aeon's `LOF` = unsupervised
- however, calling `fit` and `decision_function` corresponds to semi-supervised usage

Please have two separate tests. If the results still do not match, we need to take a deeper look.

File: aeon/anomaly_detection/_lof.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+         - binary classification
+       * - missing_values
+         - False
+       * - Learning Type
+         - unsupervised or semi-supervised
+       * - python_dependencies
+         - ["pyod"]
+
+    Parameters
+    ----------
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:36]: Please add the following statement:

```
    The documentation for parameters has been adapted from the
    [PyOD documentation](https://pyod.readthedocs.io/en/latest/pyod.models.html#id117).
    Here, `X` refers to the set of sliding windows extracted from the time series
    using :func:`aeon.utils.windowing.sliding_windows` with the parameters
    ``window_size`` and ``stride``. The internal `X` has the shape
    `(n_windows, window_size * n_channels)`.
```

File: aeon/anomaly_detection/_lof.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    This class implement metrics-based outlier detection algorithms using the
+    Local Outlier Factor (LOF) algorithm from PyOD.
+
+    .. list-table:: Capabilities
+       :stub-columns: 1
+
+       * - Input data format
+         - univariate or multivariate
+       * - Output data format
+         - binary classification
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:27]: Is that true? It should return anomaly scores, right?

File: aeon/anomaly_detection/_lof.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        n_jobs: int = 1,
+        window_size: int = 10,
+        stride: int = 1,
+        pyod_model=None,
+    ):
+        _check_soft_dependencies(*self._tags["python_dependencies"])
+        from pyod.models.lof import LOF
+
+        # Using the pyod_model if already provided else creating a new instance
+        model = pyod_model or LOF(
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:104]: Where should the `pyod_model` come from? It is not defined in this scope, so please remove.

File: aeon/anomaly_detection/_lof.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+        )
+        self.n_neighbors = n_neighbors
+        self.algorithm = algorithm
+        self.leaf_size = leaf_size
+        self.metric = metric
+        self.p = p
+        self.metric_params = metric_params
+        self.n_jobs = n_jobs
+        self.window_size = window_size
+        self.stride = stride
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:122]: Not necessary
```suggestion
```

File: aeon/anomaly_detection/_lof.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+        super().__init__(model, window_size, stride)
+
+    def _fit(self, X: np.ndarray, y: Union[np.ndarray, None] = None) -> None:
+        # Set novelty to True for supervised learning
+        self.pyod_model.novelty = True
+        super()._fit(X, y)
+        self.is_fitted = True  # Fitting completed
+        return self
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:130]: The super classes take care of this; remove.
```suggestion
```

File: aeon/anomaly_detection/_lof.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+    def _predict(self, X: np.ndarray) -> np.ndarray:
+        if not hasattr(self, "fitted_pyod_model_"):
+            raise NotFittedError(
+                "This instance of LOF has not been fitted yet; please call `fit` first."
+            )
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:136]: This should also be handled by the `BaseAnomalyDetector`-class.
```suggestion
```

File: aeon/anomaly_detection/_lof.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+        # Set novelty to True for prediction on unseen data
+        self.pyod_model.novelty = True
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:138]: Has no effect because the `PyODAdapter` uses a copy (`fitted_pyod_model_`) internally. Is not necessary either because **`_fit`** and `_fit_predict` set it already.
```suggestion
```

**Review by CodeLionX — CHANGES_REQUESTED** [review]

File: aeon/anomaly_detection/_lof.py (hunk 12 above)
Comment (aryanpola) [reply aeon/anomaly_detection/_lof.py:27]: I think it's binary when the novelty=false (unsupervised) and returns anomaly scores when novelty=true (semi-supervised).

Comment (CodeLionX) [reply aeon/anomaly_detection/_lof.py:27]: Nope, both cases return anomaly scores:

```python
>>> import numpy as np
>>> from pyod.models.lof import LOF
>>> X_train = np.random.random((20, 10))
>>> X_test = np.random.random((5, 10))
>>> unsupervised_lof = LOF(novelty=False).fit(X_train)
>>> unsupervised_lof.decision_scores_
array([1.00537546, 0.99799013, 0.99987937, 0.98546212, 1.00148011,
       1.01117638, 0.98546212, 1.00241542, 0.99992492, 0.99799013,
       0.99809342, 1.00128371, 0.99665617, 0.99972901, 0.99762421,
       1.00798589, 1.00472376, 1.0027196 , 1.00820678, 0.99657257])
>>> semi_supervised_lof = LOF(novelty=True).fit(X_train)
>>> semi_supervised_lof.decision_function(X_test)
array([0.98546212, 0.99972901, 1.00128371, 0.98546212, 0.98546212])
```

**Comment by aryanpola** [comment]
@CodeLionX Hey! I am unable to resolve these conflicts. I have tried everything, but I am stuck. Please tell me how to proceed from here.

File: aeon/anomaly_detection/tests/test_lof.py [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+        n_neighbors=n_neighbors,
+        algorithm=algorithm,
+        leaf_size=leaf_size,
+        metric=metric,
+        p=p,
+        n_jobs=n_jobs,
+        window_size=window_size,
+        stride=stride,
+    )
+    scores_aeon = lof_aeon._fit_predict(series)
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:42]: Always use the original methods!
```suggestion
    scores_aeon = lof_aeon.fit_predict(series)
```

File: aeon/anomaly_detection/tests/test_lof.py [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:148]:
```suggestion
    scores_aeon = lof_aeon.fit_predict(series)
```

File: aeon/anomaly_detection/tests/test_lof.py [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+    # Compare anomaly scores using assert_allclose
+    np.testing.assert_allclose(scores_aeon, scores_pyod, rtol=1e-5, atol=1e-5)
+
+    # Ensure that the most anomalous point is within the introduced anomaly range
+    assert (
+        50 <= np.argmax(scores_aeon) <= 58
+    ), "AEON LOF did not detect anomalies in the expected range."
+    assert (
+        50 <= np.argmax(scores_pyod) <= 58
+    ), "PyOD LOF did not detect anomalies in the expected range."
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:180]: Not necessary, please remove.

File: aeon/anomaly_detection/tests/test_lof.py [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+        algorithm=algorithm,
+        leaf_size=leaf_size,
+        metric=metric,
+        p=p,
+        n_jobs=n_jobs,
+        window_size=window_size,
+        stride=stride,
+    )
+    lof_aeon.fit(series_train)
+    scores_aeon = lof_aeon._predict(series_test)  # changed from decsion_function
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:218]:
```suggestion
    scores_aeon = lof_aeon.predict(series_test)  # changed from decsion_function
```

File: aeon/anomaly_detection/tests/test_lof.py [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+    # Compare anomaly scores using assert_allclose
+    np.testing.assert_allclose(scores_aeon, scores_pyod, rtol=1e-5, atol=1e-5)
+
+    # Ensure that the most anomalous point is within the introduced anomaly range
+    assert (
+        5 <= np.argmax(scores_aeon) <= 15
+    ), "AEON LOF did not detect anomalies in the expected range."
+    assert (
+        5 <= np.argmax(scores_pyod) <= 15
+    ), "PyOD LOF did not detect anomalies in the expected range."
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:250]: Not necessary, please remove

File: aeon/anomaly_detection/tests/test_lof.py [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+from aeon.anomaly_detection import LOF
+from aeon.testing.data_generation import make_example_1d_numpy
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+
+@pytest.mark.skipif(
+    not _check_soft_dependencies("pyod", severity="none"),
+    reason="required soft dependency PyOD not available",
+)
+def test_lof_default():
```
Comment (CodeLionX) [inline aeon/anomaly_detection/tests/test_lof.py:15]: This test looks very similar to `test_lof_unsupervised`, which already tests that our LOF has the same results as PyOD's LOF.

Please use this test to check that our LOF produces the correct output (detects the anomaly): You can remove PyOD's LOF and the comparison of the two.

**Review by CodeLionX — CHANGES_REQUESTED** [review]

File: aeon/anomaly_detection/_lof.py [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+
+    def _fit_predict(
+        self, X: np.ndarray, y: Union[np.ndarray, None] = None
+    ) -> np.ndarray:
+        # Set novelty to False for unsupervised learning
+        self.pyod_model.novelty = False
+        return super()._fit_predict(X, y)
+
+    @classmethod
+    def get_test_params(cls, parameter_set="default"):
```
Comment (CodeLionX) [inline aeon/anomaly_detection/_lof.py:147]: This is responsible for the test failures:

```suggestion
    @classmethod
    def _get_test_params(cls, parameter_set="default"):
```

**Review by CodeLionX — APPROVED** [review]

**Comment by MatthewMiddlehurst** [comment]
Seems to be a conflict here. Not sure if the unresolved conflict needs any action or not.

**Comment by SebastianSchmidl** [comment]
> Seems to be a conflict here. Not sure if the unresolved conflict needs any action or not.

Fixed.

Were just some duplicated entries in `__all__`, still ready to merge from my side.

**Review by SebastianSchmidl — APPROVED** [review]
